pkg/webrtcclient/client.go

```diff
@@ -67,66 +68,93 @@ func (c *Controller) connect(onComplete func(error)) {
+				panic(fmt.Sprintf("Failed to reshape json: %v", err))
+			}
+			if err := c.onICECandidate(candidate); err != nil {
+				return fmt.Errorf("Error handling ICE candidate: %w", err)
 			}
-			switch msg["type"] {
-			case RequestOfferMsgType:
-				panic("Device requested an offer. This violates the signaling protocol")
-			case OfferMsgType:
-				offer, err := Reshape[webrtc.SessionDescription](msg)
-				if err != nil {
-					// msg was obtained from a JSON string so Reshape shouldn't fail
-					panic(fmt.Sprintf("Failed to reshape json: %v", err))
-				}
-				if err := c.onOffer(offer); err != nil {
-					return fmt.Errorf("Error handling offer: %w", err)
-				}
-			case ICECandidateMsgType:
-				candidate, err := Reshape[webrtc.ICECandidateInit](msg)
-				if err != nil {
-					// msg was obtained from a JSON string so Reshape shouldn't fail
-					panic(fmt.Sprintf("Failed to reshape json: %v", err))
-				}
-				if err := c.onICECandidate(candidate); err != nil {
-					return fmt.Errorf("Error handling ICE candidate: %w", err)
-				}
-			case AnswerMsgType:
-				answer, err := Reshape[webrtc.SessionDescription](msg)
-				if err != nil {
-					// msg was obtained from a JSON string so Reshape shouldn't fail
-					panic(fmt.Sprintf("Failed to reshape json: %v", err))
-				}
-				if err := c.onAnswer(answer); err != nil {
-					return fmt.Errorf("Error handling answer: %w", err)
-				}
-			default:
-				if errMsg, errPresent := msg["error"]; errPresent {
-					return fmt.Errorf("Received error from signaling server: %v", errMsg)
-				} else {
-					return fmt.Errorf("Unknown message type: %v", msg["type"])
-				}
+		case AnswerMsgType:
+			answer, err := Reshape[webrtc.SessionDescription](msg)
+			if err != nil {
+				// msg was obtained from a JSON string so Reshape shouldn't fail
+				panic(fmt.Sprintf("Failed to reshape json: %v", err))
+			}
+			if err := c.onAnswer(answer); err != nil {
+				return fmt.Errorf("Error handling answer: %w", err)
+			}
+		default:
+			if errMsg, errPresent := msg["error"]; errPresent {
+				return fmt.Errorf("Received error from signaling server: %v", errMsg)
+			} else {
+				return fmt.Errorf("Unknown message type: %v", msg["type"])
 			}
 		}
 	}
 }
 
+func (c *Controller) stopGoRoutines() {
+	// When the peer connection closes the signaling goroutines must be stopped.
+	// This process is triggered by closing the send channel which will notify
+	// the forwarding signaling goroutine to exit since it won't have nothing
+	// else to forward. This will eventually lead to the recv channel to close,
+	// which will finally cause the go routine running the recv loop to exit and
+	// complete the shutdown mechanism.
+	c.closeSendCh()
+}
+
+func (c *Controller) closeSendCh() {
+	c.sendMtx.Lock()
+	defer c.sendMtx.Unlock()
+	if !c.sendChClosed {
+		close(c.signaling.SendCh)
+		c.sendChClosed = true
+	}
+}
+
+func (c *Controller) sendSigMsg(msg interface{}) {
```

Comment:
The problem with that approach is that the channel could be closed after the check returns `false` but before the goroutine has a chance to `send` on the channel, which leads to a panic.

Ideally, I'd like to close the channel from the same goroutine that writes to it, for which your extra channel (or an atomic variable) would be useful. But in this case writes to the channel occur from many goroutines (created by the webrtc library, so there is nothing I can do about it). I need a way to avoid the channel closing between the check and the send, for which a mutex is the best choice.

I thought about using a RWLock instead to allow several routines to `send` on the channel at the same time, but in practice the channel won't allow this and instead using its own blocking mechanism to ensure only one sends at a time. The other downside of a RWLock is that several routines trying to "read" (i.e `send`) could starve a single routine trying to "write" (i.e close the channel), there are strategies to avoid this too, but at that point is some extra complication that is simply not worth it: The channel is written to relatively rarely so a regular mutex & non buffered channel should have no performance impact.